addons/sale_timesheet/models/account_invoice.py

```diff
@@ -19,16 +20,43 @@ def invoice_validate(self):
                     ('timesheet_invoice_id', '=', False),
                     ('timesheet_invoice_type', 'in', ['billable_time', 'billable_fixed'])
                 ])
+                if uninvoiced_timesheet_lines:
+                    precision = invoice_line.currency_id.decimal_places
+                    # NOTE JEM : changing quantity (or unit price) of invoice line does not impact the revenue calculation. (FP specs)
+                    # delivered : update revenue with the prorata of number of hours on the timesheet line
+                    if invoice_line.product_id.invoice_policy == 'delivery':
+                        invoiced_price_per_hour = float_round(invoice_line.price_subtotal_signed / float(sum(uninvoiced_timesheet_lines.mapped('unit_amount'))), precision)
+                        total_revenue = 0.0
+                        for timesheet_line in uninvoiced_timesheet_lines.with_context(create=True)[:-1]:
+                            line_revenue = float_round(invoiced_price_per_hour * timesheet_line.unit_amount)
+                            total_revenue += line_revenue
+                            timesheet_line.write({
+                                'timesheet_invoice_id': invoice.id,
+                                'timesheet_revenue': line_revenue,
+                            })
+                        # avoid rounding problem : the last line contains the rest to make the total right
+                        uninvoiced_timesheet_lines[-1].with_context(create=True).write({
```

Comment:
why not just setting the context on `uninvoiced_timesheet_lines` (or an alias) then fucking around with that, so that you don't keep blowing your env and cache?